src/Diagnostics/Samples/CSharp/CSharpAnalyzers/CSharpAnalyzers/StatefulAnalyzers/CodeBlockStartedAnalyzer.cs

```diff
@@ -0,0 +1,128 @@
+    /// <para>
+    /// The analyzer registers:
+    /// (a) A code block start action, which initializes per-code block mutable state. We mark all parameters as unused at start of analysis.
+    /// (b) A code block syntax node action, which identifes parameter references and marks the corresponding parameter as used.
+    /// (c) A code block end action, which reports diagnostics based on the final state, for all parameters which are unused.
+    /// </para>
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class CodeBlockStartedAnalyzer : DiagnosticAnalyzer
+    {
+        #region Descriptor fields
+
+        internal static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.CodeBlockStartedAnalyzerTitle), Resources.ResourceManager, typeof(Resources));
+        internal static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.CodeBlockStartedAnalyzerMessageFormat), Resources.ResourceManager, typeof(Resources));
+        internal static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.CodeBlockStartedAnalyzerDescription), Resources.ResourceManager, typeof(Resources));
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticIds.CodeBlockStartedAnalyzerRuleId, Title, MessageFormat, DiagnosticCategories.Stateful, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        #endregion
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCodeBlockStartAction<SyntaxKind>(startCodeBlockContext =>
+            {
+                // We only care about method bodies.
+                if (startCodeBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                {
+                    return;
+                }
+
+                // We only care about methods with parameters.
+                var method = (IMethodSymbol)startCodeBlockContext.OwningSymbol;
+                if (method.Parameters.IsEmpty)
+                {
+                    return;
+                }
+
+                // Initialize local mutable state in the start action.
+                var analyzer = new UnusedParametersAnalyzer(method);
+
+                // Register an intermediate non-end action that accesses and modifies the state.
+                startCodeBlockContext.RegisterSyntaxNodeAction(analyzer.AnalyzeSyntaxNode, SyntaxKind.IdentifierName);
+
+                // Register an end action to report diagnostics based on the final state.
+                startCodeBlockContext.RegisterCodeBlockEndAction(analyzer.CodeBlockEndAction);
+            });
+        }
+
+        private class UnusedParametersAnalyzer
```

Comment:
I am really scared of promoting use of locals within Initialize methods or inner lambdas, especially for mutable state. One might end up thinking what harm to just move them to the top of the method body and you inadvertently introduce a performance bottleneck.

Also, I feel having all analysis code in lambda might make sense for very simple analyzers, but makes it confusing if that code grows, making it difficult to read and find just the registered actions. I think `Initialize` method's body should just deal with registering all actions/callbacks, with lambdas used for registering nested actions and analysis should live in separate methods. Probably that's my preference, but if you feel strongly I can change the CompilationStartedAnalyzer's field (which is immutable state) to a local.